ipnsfs/system.go

```diff
@@ -0,0 +1,285 @@
+	if err != nil {
+		err = namesys.InitializeKeyspace(ctx, fs.dserv, fs.nsys, fs.pins, k)
+		if err != nil {
+			return nil, err
+		}
+
+		pointsTo, err = fs.nsys.Resolve(ctx, name)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	mnode, err := fs.dserv.Get(pointsTo)
+	if err != nil {
+		return nil, err
+	}
+
+	root.node = mnode
+
+	root.repub = NewRepublisher(root, time.Millisecond*300, time.Second*3)
+	go root.repub.Run(parent)
+
+	pbn, err := ft.FromBytes(mnode.Data)
+	if err != nil {
+		log.Error("IPNS pointer was not unixfs node")
+		return nil, err
+	}
+
+	switch pbn.GetType() {
+	case ft.TDirectory:
+		root.val = NewDirectory(pointsTo.B58String(), mnode, root, fs)
+	case ft.TFile, ft.TMetadata, ft.TRaw:
+		fi, err := NewFile(pointsTo.B58String(), mnode, root, fs)
+		if err != nil {
+			return nil, err
+		}
+		root.val = fi
+	default:
+		panic("unrecognized! (NYI)")
+	}
+	return root, nil
+}
+
+func (kr *KeyRoot) GetValue() FSNode {
+	return kr.val
+}
+
+func (kr *KeyRoot) Open(tpath []string, mode int) (File, error) {
+	if kr.val == nil {
+		// No entry here... what should we do?
+		panic("nyi")
+	}
+	if len(tpath) > 0 {
+		// Make sure our root is a directory
+		dir, ok := kr.val.(*Directory)
+		if !ok {
+			return nil, fmt.Errorf("no such file or directory: %s", tpath[0])
+		}
+
+		return dir.Open(tpath, mode)
+	}
+
+	switch t := kr.val.(type) {
+	case *Directory:
+		return nil, ErrIsDirectory
+	case File:
+		return t, nil
+	default:
+		panic("unrecognized type, should not happen")
+	}
+}
+
+// closeChild implements the childCloser interface, and signals to the publisher that
+// there are changes ready to be published
+func (kr *KeyRoot) closeChild(name string, nd *dag.Node) error {
+	kr.repub.Touch()
+	return nil
+}
+
+// Publish publishes the ipns entry associated with this key
+func (kr *KeyRoot) Publish(ctx context.Context) error {
+	child, ok := kr.val.(FSNode)
+	if !ok {
+		return errors.New("child of key root not valid type")
+	}
+
+	nd, err := child.GetNode()
+	if err != nil {
+		return err
+	}
+
+	k, err := kr.fs.dserv.Add(nd)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Publishing!")
+	return kr.fs.nsys.Publish(ctx, kr.key, k)
+}
+
+// Republisher manages when to publish the ipns entry associated with a given key
+type Republisher struct {
+	TimeoutLong  time.Duration
+	TimeoutShort time.Duration
+	Publish      chan struct{}
+	root         *KeyRoot
+}
+
+func NewRepublisher(root *KeyRoot, tshort, tlong time.Duration) *Republisher {
+	return &Republisher{
+		TimeoutShort: tshort,
+		TimeoutLong:  tlong,
+		Publish:      make(chan struct{}, 1),
+		root:         root,
+	}
+}
+
+func (np *Republisher) Touch() {
+	select {
+	case np.Publish <- struct{}{}:
+	default:
+	}
+}
+
+func (np *Republisher) Run(ctx context.Context) {
+	for {
+		select {
+		case <-np.Publish:
+			quick := time.After(np.TimeoutShort)
+			longer := time.After(np.TimeoutLong)
+
+		wait:
+			select {
+			case <-quick:
+			case <-longer:
```

Comment:
maybe this order?

```go
select {
case <-ctx.Done():
  ...
case <-np.Publish:
  ...
case <-quick:
case <-longer:
}
```
so it's clear the last two just fall through out of the select.  when first reading this i somehow took it to mean all three (quick, longer, ctx.Done) returned-- which is not correct Go, but the code sort of leads me there